pkg/kubelet/nodestatus/setters.go

```diff
@@ -192,6 +174,55 @@ func NodeAddress(nodeIP net.IP, // typically Kubelet.nodeIP
 	}
 }
 
+// shouldUseCloudProviderReportedHostname decides whether the given hostname or
+// the one reported by the cloud provider should be used
+func shouldUseCloudProviderReportedHostname(hostname string, // typically Kubelet.hostname, reported by the OS
+	hostnameOverridden bool,
+	cloudNodeAddresses []v1.NodeAddress) bool {
+	switch {
+	case len(cloudNodeAddresses) == 0:
+		// the cloud provider didn't specify any addresses
+		return false
+
+	case !hasAddressType(cloudNodeAddresses, v1.NodeHostName) && hasAddressValue(cloudNodeAddresses, hostname):
+		// the cloud provider didn't specify an address of type Hostname,
+		// but the auto-detected hostname matched an address reported by the cloud provider,
+		// so we can add it and count on the value being verifiable via cloud provider metadata
+		return false
+
+	case hostnameOverridden:
+		// the hostname was force-set via flag/config.
+		// this means the hostname might not be able to be validated via cloud provider metadata,
+		// but was a choice by the kubelet deployer we should honor
+		return false
+	}
+
+	return true
+}
+
+// ensureHostname ensures the given hostname is present in the given nodeAddresses
+func ensureHostname(hostname string, nodeAddresses []v1.NodeAddress) []v1.NodeAddress {
+	var existingHostnameAddress *v1.NodeAddress
+	for i := range nodeAddresses {
+		if nodeAddresses[i].Type == v1.NodeHostName {
+			existingHostnameAddress = &nodeAddresses[i]
+			break
+		}
+	}
+
+	if existingHostnameAddress == nil {
+		// no existing Hostname address found, add it
+		klog.Warningf("adding overridden hostname of %v to cloudprovider-reported addresses", hostname)
```

Comment:
this warns unnecessarily in the case where the cloud-provider reported an address that matches this value, but just didn't happen to be of Hostname-type